File: entry/src/main/js/MainAbility/health-monitoring/domain/baseline.js
import { err, ok } from './model.js';

export function median(values) {
    if (!values || values.length === 0) {
        return null;
    }
    const items = values.slice().sort(function (a, b) { return a - b; });
    const middle = Math.floor(items.length / 2);
    return items.length % 2 === 0 ? (items[middle - 1] + items[middle]) / 2 : items[middle];
}

export function mean(values) {
    if (!values || values.length === 0) {
        return null;
    }
    let total = 0;
    for (let index = 0; index < values.length; index += 1) {
        total += values[index];
    }
    return total / values.length;
}

export function mad(values, center) {
    const base = center === undefined || center === null ? median(values) : center;
    if (base === null) {
        return null;
    }
    return median(values.map(function (value) { return Math.abs(value - base); }));
}

export function buildPersonalBaseline(metricId, metrics, options) {
    const rules = options || {};
    const accepted = (metrics || []).filter(function (item) {
        return item.metricId === metricId && item.quality && item.quality.tag !== 'Rejected';
    });
    const minimum = rules.minimumSamples === undefined ? 7 : rules.minimumSamples;
    if (accepted.length < minimum) {
        return err('INSUFFICIENT_BASELINE', { required: minimum, actual: accepted.length });
    }
    const values = accepted.map(function (item) { return item.value; });
    const center = median(values);
    return ok(Object.freeze({
        tag: 'PersonalBaseline',
        metricId: metricId,
        context: rules.context || 'default',
        sampleCount: accepted.length,
        median: center,
        mad: mad(values, center),
        algorithmVersion: 'robust-median-mad/1.0.0',
        inputMetricIds: Object.freeze(accepted.map(function (item) { return item.id; }))
    }));
}

export function compareToBaseline(baseline, metric) {
    if (!baseline || !metric || baseline.metricId !== metric.metricId) {
        return err('BASELINE_METRIC_MISMATCH');
    }
    const scale = baseline.mad && baseline.mad > 0 ? baseline.mad * 1.4826 : null;
    const robustZ = scale ? (metric.value - baseline.median) / scale : 0;
    return ok(Object.freeze({
        tag: 'Deviation',
        metricId: metric.metricId,
        value: metric.value,
        baseline: baseline.median,
        robustZ: robustZ,
        direction: robustZ > 0.5 ? 'up' : (robustZ < -0.5 ? 'down' : 'stable'),
        unusual: Math.abs(robustZ) >= 3,
        medicalMeaning: 'not_assessed'
    }));
}

File: health-android16/domain/src/main/kotlin/com/move25/health/domain/Baseline.kt
package com.move25.health.domain

import kotlin.math.abs

data class PersonalBaseline(
    val subjectId: SubjectId,
    val metricId: MetricId,
    val interval: TimeInterval,
    val median: Double,
    val mad: Double,
    val sampleCount: Int,
    val inputMetricIds: List<String>,
    val version: String = "robust-baseline/1.0.0",
)

data class Deviation(
    val metricId: MetricId,
    val observed: Double,
    val baseline: Double,
    val absoluteDelta: Double,
    val percentageDelta: Double?,
    val robustZ: Double?,
    val unusual: Boolean,
    val direction: String,
)

fun buildBaseline(metrics: List<DerivedMetric>, minimumSamples: Int = 7): Result<DomainError, PersonalBaseline> {
    val qualified = metrics.filter { it.quality !is DataQuality.Rejected }
    if (qualified.size < minimumSamples) return Result.Err(DomainError("INSUFFICIENT_BASELINE_HISTORY"))
    if (qualified.map { it.subjectId }.distinct().size != 1 || qualified.map { it.metricId }.distinct().size != 1) {
        return Result.Err(DomainError("MIXED_BASELINE_SERIES"))
    }
    val center = median(qualified.map { it.value })
    val mad = median(qualified.map { abs(it.value - center) })
    val intervalResult = TimeInterval.of(
        qualified.minOf { it.interval.start.value },
        qualified.maxOf { it.interval.endExclusive.value },
    )
    if (intervalResult is Result.Err) return intervalResult
    val interval = intervalResult.getOrNull() ?: return Result.Err(DomainError("BASELINE_INTERVAL_INVALID"))
    return Result.Ok(PersonalBaseline(
        qualified.first().subjectId, qualified.first().metricId, interval, center, mad,
        qualified.size, qualified.map { it.id },
    ))
}

fun compareToBaseline(baseline: PersonalBaseline, metric: DerivedMetric, threshold: Double = 3.0): Result<DomainError, Deviation> {
    if (baseline.subjectId != metric.subjectId || baseline.metricId != metric.metricId) return Result.Err(DomainError("BASELINE_MISMATCH"))
    val delta = metric.value - baseline.median
    val robustZ = if (baseline.mad > 0) delta / (baseline.mad * 1.4826) else null
    return Result.Ok(Deviation(
        metric.metricId, metric.value, baseline.median, delta,
        baseline.median.takeIf { it != 0.0 }?.let { delta / it * 100.0 }, robustZ,
        robustZ?.let { abs(it) >= threshold } ?: false,
        when { delta > 0 -> "up"; delta < 0 -> "down"; else -> "stable" },
    ))
}

data class PersistentChange(
    val changed: Boolean,
    val direction: String,
    val persistence: Int,
    val robustScores: List<Double>,
    val medicalMeaning: String = "not_assessed",
)

fun detectPersistentChange(metrics: List<DerivedMetric>, baseline: PersonalBaseline, required: Int = 3): Result<DomainError, PersistentChange> {
    if (metrics.size < required || baseline.mad <= 0) return Result.Err(DomainError("INSUFFICIENT_CHANGE_EVIDENCE"))
    val scores = metrics.takeLast(required).map { (it.value - baseline.median) / (baseline.mad * 1.4826) }
    val up = scores.all { it >= 3.0 }
    val down = scores.all { it <= -3.0 }
    return Result.Ok(PersistentChange(up || down, if (up) "up" else if (down) "down" else "variable", required, scores))
}
